library/core/src/intrinsics.rs

```diff
@@ -2470,6 +2470,30 @@ extern "rust-intrinsic" {
     /// constructing an empty slice) is returned.
     #[rustc_nounwind]
     pub fn option_payload_ptr<T>(arg: *const Option<T>) -> *const T;
+
+    /// Returns whether the argument is known at compile-time.
+    ///
+    /// This is useful when there is a way of writing the code that will
+    /// be *faster* when some variables have known values, but *slower*
+    /// in the general case: an `if is_compile_time_known(var)` can be used
+    /// to select between these two variants. The `if` will be optimized away
+    /// and only the desired branch remains.
+    ///
+    /// Formally speaking, this function non-deterministically returns `true`
+    /// or `false`, and the caller has to ensure sound behavior for both cases.
+    /// In other words, the following code has *Undefined Behavior*:
+    ///
+    /// ```rust
+    /// if !is_compile_time_known(0) { unreachable_unchecked(); }
+    /// ```
+    ///
+    /// Unsafe code may not rely on `is_compile_time_known` returning any
+    /// particular value, ever. However, the compiler will generally make it
+    /// return `true` only if the value of the argument is actually known.
```

Comment:
This needs an added paragraph on the const caveats. When calling this inside a `const fn`, you must ensure that both branches are semantically equivalent, similar to `const_eval_select`.